include/vkwrap/vkwrap/mman.h

```diff
@@ -0,0 +1,538 @@
+         */
+        .srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED,
+        .dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED,
+
+        .image = image,
+
+        .subresourceRange = {
+            .aspectMask = chooseAspectMask( format, new_layout ),
+
+            // We use images with one layer and one mipmap.
+            .baseMipLevel = 0,
+            .levelCount = 1,
+            .baseArrayLayer = 0,
+            .layerCount = 1,
+        }
+    };
+    // clang-format on
+
+    return barrier;
+} // createImageBarrierInfo
+
+inline PipelineStages
+choosePipelineStages( vk::ImageLayout old_layout, vk::ImageLayout new_layout )
+{
+    using vk::ImageLayout;
+    using vk::PipelineStageFlagBits;
+    using vk::PipelineStageFlags;
+
+    PipelineStageFlags src{ PipelineStageFlagBits::eNone };
+    PipelineStageFlags dst{ PipelineStageFlagBits::eNone };
+
+    if ( old_layout == ImageLayout::eUndefined && new_layout == ImageLayout::eTransferDstOptimal )
+    {
+        src = PipelineStageFlagBits::eTopOfPipe;
+        dst = PipelineStageFlagBits::eTransfer;
+    } else if ( old_layout == ImageLayout::eTransferDstOptimal && new_layout == ImageLayout::eShaderReadOnlyOptimal )
+    {
+        src = PipelineStageFlagBits::eTransfer;
+        dst = PipelineStageFlagBits::eFragmentShader;
+    } else if ( old_layout == ImageLayout::eUndefined && new_layout == ImageLayout::eDepthStencilAttachmentOptimal )
+    {
+        src = PipelineStageFlagBits::eTopOfPipe;
+        dst = PipelineStageFlagBits::eEarlyFragmentTests;
+    } else
+    {
+        throw Error{ "choosePipelineStages: unimplemented layout transition." };
+    }
+
+    return { src, dst };
+} // choosePipelineStages
+
+class Mman
+{
+
+  private:
+    struct BufferInfo
+    {
+        VmaAllocation allocation;
+        vk::DeviceSize size;
+    }; // struct BufferInfo
+
+    struct ImageInfo
+    {
+        VmaAllocation allocation;
+        vk::Format format;
+        vk::ImageLayout layout;
+        vk::Extent3D extent;
+    }; // struct ImageInfo
+
+    class OneTimeCommand : private vk::UniqueCommandBuffer
+    {
+
+      private:
+        using Base = vk::UniqueCommandBuffer;
+
+      private:
+        vk::Queue m_queue;
+
+        void begin() const
+        {
+            // clang-format off
+            vk::CommandBufferBeginInfo begin_info{ 
+                .flags = vk::CommandBufferUsageFlagBits::eOneTimeSubmit };
+            // clang-format on
+
+            Base::get().begin( begin_info );
+        } // begin
+
+        void end() const { Base::get().end(); }
+
+        void submit() const
+        {
+            vk::SubmitInfo submit_info{ .commandBufferCount = 1, .pCommandBuffers = &Base::get() };
+
+            m_queue.submit( std::array{ submit_info } );
+        } // submit
+
+        void wait() const { m_queue.waitIdle(); }
+
+      public:
+        OneTimeCommand( vk::UniqueCommandBuffer&& unique_cmd, vk::Queue queue )
```

Comment:
You can remove && rvalue reference, because you anyway use `std::move`  for unique_cmd and `vk::UniqueCommandBuffer` has deleted copy constructor and assignment

Reply:
Agree.